src/module_4/fit_model.py

```
@@ -0,0 +1,102 @@
+
+
+def save_model(model) -> Path:
+    """Saves the trained GradientBoostingTree model on disk with joblib extension
+
+    Args:
+        - model: fitted model
+
+    Raises:
+        FileExistsError: if the fie path of the model already exists
+
+    Returns:
+        file_path = path to the stored model (type: model_name_pushh_yyyy_mm_dd where yyyy_mm_dd is the training date)
+    """
+
+    file_path = (
+        Path(__file__).parent.parent.parent.parent.resolve()
+        / f"models/{create_model_file_name('GradientBoostingTree')}"
+    )
+
+    if file_path.exists():
+        raise FileExistsError(f"{file_path} alrady exists")
+
+    joblib.dump(model, file_path)
+    return file_path
+
+
+def create_model_file_name(model_name: str) -> str:
+    actual_date = datetime.today().strftime("%Y-%m-%d")
+    model_name = f"GradientBoostingTree_push_{actual_date}.joblib"
+    return model_name
+
+
+def handler_fit(event: dict, _) -> dict[str, any]:
+    """Fits the model with GradientBoostingTree
+
+    Args:
+        - event: dictionary with model parameters:
+        (
+            "model_parameters":
+            {
+                "GradientBoosting param1": value,
+                "GradientBoosting param2": value,
+                ....
+            }
+            "model_threshold": value
+        )
+        The model_threshold is optional, only model_parameters is mandatory
+
+        Example:
+            (
+                "model_parameters":
+                {
+                    "learning_rate": 0.1,
+                    "n_estimators": 100,
+                }
+            )
+
+    Returns:
+        response: Dictionary with "statusCode" (200 if OK; 500 if ERR) and "body" with the "model_path": value, where the model has been saved
+    """
+
+    model_parameters, prediction_threshold = extract_parameters(event)
+    model = PushModel(model_parameters, prediction_threshold)
+
+    trainnig_data = load_clear_data()
+    model.fit(trainnig_data)
```

Subtle thing, if you check if the path exists you can raise the exception earlier and avoid time and resources training a model that will not be saved

---

Okey